apm-agent-core/src/main/java/co/elastic/apm/agent/report/ssl/TLSFallbackSSLSocket.java

```diff
@@ -0,0 +1,408 @@
+/*-
+ * #%L
+ * Elastic APM Java agent
+ * %%
+ * Copyright (C) 2018 - 2020 Elastic and contributors
+ * %%
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * #L%
+ */
+package co.elastic.apm.agent.report.ssl;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.net.ssl.HandshakeCompletedListener;
+import javax.net.ssl.SSLHandshakeException;
+import javax.net.ssl.SSLParameters;
+import javax.net.ssl.SSLSession;
+import javax.net.ssl.SSLSocket;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.InetAddress;
+import java.net.Socket;
+import java.net.SocketAddress;
+import java.net.SocketException;
+import java.net.SocketImplFactory;
+import java.nio.channels.SocketChannel;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
+
+class TLSFallbackSSLSocket extends SSLSocket {
+
+    private static final Logger logger = LoggerFactory.getLogger(TLSFallbackSSLSocket.class);
+
+    private static final String TLS_v_1_3 = "TLSv1.3";
+
+    private final TLSFallbackSSLSocketFactory socketFactory;
+
+    private SSLSocket socket;
+
+    TLSFallbackSSLSocket(SSLSocket socket, TLSFallbackSSLSocketFactory socketFactory) {
+        this.socket = socket;
+        this.socketFactory = socketFactory;
+    }
+
+    @Override
+    public void startHandshake() throws IOException {
+
+        // known JDK bug: JDK-8236039
+        // automatically fallback to another protocol when triggered
+        try {
+            socket.startHandshake();
+        } catch (SSLHandshakeException e) {
+
+            Set<String> enabledProtocols = new HashSet<>(Arrays.asList(socket.getEnabledProtocols()));
+            if (enabledProtocols.contains(TLS_v_1_3) && e.getMessage().contains("should not be presented in")) {
+
+                logger.warn("Workaround for JDK Bug JDK-8236039 applied, will connect without TLS v1.3");
```

Comment:
Seems like we're re-triggering this exception for every request to APM Server, right? Could we exclude TLSv1.3 if we know it triggers the bug on the current JVM?

Reply:
Maybe also suggest the user what to do when the exception occurs, like updating the JDK, or removing 1.3 for the server's setting (https://www.elastic.co/guide/en/apm/server/current/agent-server-ssl.html#_supported_protocols_2).

Reply:
I agree with both comments, having  a single warning is better, and an actionable message is definitely better than having to deal with that in an FAQ somewhere :-).

Reply:
Also, there would be the option to disable TLS 1.3 client side, but that would impact the whole JVM, or provide a configuration for agent, which implies some extra complexity for documentation/support.